File: backend/schemas/analytics.py
"""
Analytics Schemas
Pydantic models for analytics and reporting API requests and responses
"""

from pydantic import BaseModel, validator
from typing import Optional, List, Dict, Any, Union
from datetime import datetime, date
from enum import Enum

class ReportType(str, Enum):
    """Report type options"""
    REVENUE = "revenue"
    OCCUPANCY = "occupancy"
    GUEST = "guest"
    PERFORMANCE = "performance"
    COMPETITIVE = "competitive"
    FORECAST = "forecast"
    CUSTOM = "custom"

class GroupBy(str, Enum):
    """Grouping options for analytics"""
    DAY = "day"
    WEEK = "week"
    MONTH = "month"
    QUARTER = "quarter"
    YEAR = "year"

class MetricType(str, Enum):
    """Metric type options"""
    REVENUE = "revenue"
    OCCUPANCY = "occupancy"
    ADR = "adr"
    REVPAR = "revpar"
    GUEST_COUNT = "guest_count"
    BOOKING_COUNT = "booking_count"
    SATISFACTION = "satisfaction"

# Base analytics request schema
class AnalyticsRequest(BaseModel):
    """Base schema for analytics requests"""
    property_id: str
    start_date: datetime
    end_date: datetime
    group_by: GroupBy = GroupBy.DAY

# Revenue analytics schema
class RevenueAnalyticsRequest(AnalyticsRequest):
    """Schema for revenue analytics request"""
    include_breakdown: bool = True
    include_growth: bool = True

class RevenueAnalyticsResponse(BaseModel):
    """Schema for revenue analytics response"""
    property_id: str
    period: Dict[str, str]
    summary: Dict[str, float]
    revenue_trend: List[Dict[str, Any]]
    room_type_breakdown: List[Dict[str, Any]]
    generated_at: str

# Occupancy analytics schema
class OccupancyAnalyticsRequest(AnalyticsRequest):
    """Schema for occupancy analytics request"""
    include_room_types: bool = True
    include_seasonality: bool = False

class OccupancyAnalyticsResponse(BaseModel):
    """Schema for occupancy analytics response"""
    property_id: str
    period: Dict[str, str]
    summary: Dict[str, float]
    daily_occupancy: List[Dict[str, Any]]
    generated_at: str

# Guest analytics schema
class GuestAnalyticsRequest(BaseModel):
    """Schema for guest analytics request"""
    property_id: str
    days: int = 30
    include_demographics: bool = True
    include_behavior: bool = True

class GuestAnalyticsResponse(BaseModel):
    """Schema for guest analytics response"""
    property_id: str
    period_days: int
    summary: Dict[str, Any]
    demographics: Dict[str, Any]
    generated_at: str

# Performance analytics schema
class PerformanceAnalyticsRequest(BaseModel):
    """Schema for performance analytics request"""
    property_id: str
    days: int = 30
    include_trends: bool = True
    include_forecasting: bool = False

class PerformanceAnalyticsResponse(BaseModel):
    """Schema for performance analytics response"""
    property_id: str
    property_name: str
    period_days: int
    kpis: Dict[str, Any]
    revenue_metrics: Dict[str, Any]
    occupancy_metrics: Dict[str, Any]
    guest_metrics: Dict[str, Any]
    market_comparison: Dict[str, Any]
    trends: Dict[str, Any]
    generated_at: str

# Competitive analysis schema
class CompetitiveAnalysisRequest(BaseModel):
    """Schema for competitive analysis request"""
    property_id: str
    competitor_properties: List[str]
    metrics: List[MetricType] = [MetricType.OCCUPANCY, MetricType.ADR, MetricType.REVPAR]

class CompetitiveAnalysisResponse(BaseModel):
    """Schema for competitive analysis response"""
    property_id: str
    competitors: int
    analysis: Dict[str, Any]
    generated_at: str

# Forecasting schema
class ForecastingRequest(BaseModel):
    """Schema for forecasting request"""
    property_id: str
    forecast_days: int = 30
    include_confidence: bool = True
    include_scenarios: bool = False

class ForecastingResponse(BaseModel):
    """Schema for forecasting response"""
    property_id: str
    forecast_days: int
    forecast: Dict[str, Any]
    confidence_level: float
    generated_at: str

# Custom report schema
class CustomReportRequest(BaseModel):
    """Schema for custom report request"""
    property_id: str
    report_type: ReportType = ReportType.CUSTOM
    start_date: datetime
    end_date: datetime
    metrics: List[MetricType]
    filters: Optional[Dict[str, Any]] = None
    group_by: Optional[GroupBy] = None
    format: str = "json"  # json, csv, pdf

class CustomReportResponse(BaseModel):
    """Schema for custom report response"""
    property_id: str
    report_type: str
    period: Dict[str, str]
    metrics: Dict[str, Any]
    generated_at: str

# KPI schema
class KPISummary(BaseModel):
    """Schema for KPI summary"""
    total_revenue: float
    average_occupancy: float
    average_daily_rate: float
    revenue_per_available_room: float
    total_guests: int
    repeat_guest_rate: float
    average_stay_length: float
    revenue_growth: float

# Trend analysis schema
class TrendAnalysis(BaseModel):
    """Schema for trend analysis"""
    metric: str
    trend_direction: str  # increasing, decreasing, stable
    change_percentage: float
    period: str
    confidence: float

# Market comparison schema
class MarketComparison(BaseModel):
    """Schema for market comparison"""
    metric: str
    property_value: float
    market_average: float
    variance: float
    percentile: int
    performance: str  # above_average, average, below_average

# Dashboard data schema
class DashboardData(BaseModel):
    """Schema for dashboard data"""
    property_id: str
    property_name: str
    current_period: Dict[str, str]
    kpis: KPISummary
    trends: List[TrendAnalysis]
    market_comparison: List[MarketComparison]
    alerts: List[Dict[str, Any]]
    generated_at: str

# Alert schema
class AnalyticsAlert(BaseModel):
    """Schema for analytics alerts"""
    id: str
    type: str  # warning, error, info, success
    title: str
    message: str
    metric: str
    current_value: float
    threshold_value: float
    severity: str  # low, medium, high, critical
    created_at: datetime
    acknowledged: bool = False

# Export schema
class AnalyticsExport(BaseModel):
    """Schema for analytics export"""
    property_id: str
    report_type: ReportType
    start_date: datetime
    end_date: datetime
    format: str = "csv"  # csv, json, xlsx, pdf
    metrics: List[MetricType]
    filters: Optional[Dict[str, Any]] = None

# Benchmark schema
class BenchmarkData(BaseModel):
    """Schema for benchmark data"""
    property_id: str
    metric: str
    current_value: float
    benchmark_value: float
    industry_average: float
    percentile_rank: int
    performance_rating: str  # excellent, good, average, below_average, poor

# Seasonality analysis schema
class SeasonalityAnalysis(BaseModel):
    """Schema for seasonality analysis"""
    property_id: str
    metric: str
    seasonal_index: Dict[str, float]  # month -> index
    peak_season: List[str]
    low_season: List[str]
    seasonality_strength: float  # 0-1 scale
    recommendations: List[str]

# Cohort analysis schema
class CohortAnalysis(BaseModel):
    """Schema for cohort analysis"""
    property_id: str
    cohort_type: str  # booking_month, guest_type, etc.
    cohorts: List[Dict[str, Any]]
    retention_rates: Dict[str, float]
    average_lifetime_value: float
    churn_rate: float

# Real-time metrics schema
class RealTimeMetrics(BaseModel):
    """Schema for real-time metrics"""
    property_id: str
    timestamp: datetime
    current_occupancy: float
    today_revenue: float
    today_bookings: int
    pending_check_ins: int
    pending_check_outs: int
    available_rooms: int
    alerts_count: int

# Performance comparison schema
class PerformanceComparison(BaseModel):
    """Schema for performance comparison"""
    property_id: str
    comparison_period: str
    current_period: Dict[str, Any]
    previous_period: Dict[str, Any]
    changes: Dict[str, float]
    insights: List[str]

# API response schemas
class AnalyticsListResponse(BaseModel):
    """Schema for analytics list response"""
    analytics: List[Dict[str, Any]]
    total: int
    page: int
    size: int
    pages: int

class AnalyticsErrorResponse(BaseModel):
    """Schema for analytics error response"""
    success: bool = False
    error: str
    details: Optional[Dict[str, Any]] = None
    generated_at: str

# Bulk analytics schema
class BulkAnalyticsRequest(BaseModel):
    """Schema for bulk analytics request"""
    property_ids: List[str]
    report_type: ReportType
    start_date: datetime
    end_date: datetime
    metrics: List[MetricType]
    group_by: GroupBy = GroupBy.DAY

class BulkAnalyticsResponse(BaseModel):
    """Schema for bulk analytics response"""
    properties: List[Dict[str, Any]]
    summary: Dict[str, Any]
    generated_at: str

# Scheduled report schema
class ScheduledReport(BaseModel):
    """Schema for scheduled report"""
    id: str
    property_id: str
    report_type: ReportType
    schedule: str  # cron expression
    recipients: List[str]
    metrics: List[MetricType]
    is_active: bool
    created_at: datetime
    last_run: Optional[datetime] = None
    next_run: Optional[datetime] = None

# Report template schema
class ReportTemplate(BaseModel):
    """Schema for report template"""
    id: str
    name: str
    description: str
    report_type: ReportType
    metrics: List[MetricType]
    filters: Dict[str, Any]
    group_by: GroupBy
    format: str
    is_public: bool
    created_by: str
    created_at: datetime

# Analytics configuration schema
class AnalyticsConfig(BaseModel):
    """Schema for analytics configuration"""
    property_id: str
    auto_reports: bool
    alert_thresholds: Dict[str, float]
    dashboard_widgets: List[str]
    report_schedule: Optional[str] = None
    data_retention_days: int = 365
    real_time_enabled: bool = True

File: frontend/lib/types/analytics.ts
/**
 * Analytics and Business Intelligence Type Definitions
 *
 * Purpose: Type definitions for analytics, reporting, business intelligence, and forecasting
 * in Buffr Host. These types align with backend Python schemas for consistent data structures.
 * Location: lib/types/analytics.ts
 * Usage: Shared across analytics components, BI dashboards, reporting APIs, and ML forecasting services
 *
 * @module Analytics Types
 * @author Buffr Host Development Team
 * @version 1.0.0
 */

/**
 * Report Type Enumeration
 *
 * Defines the different types of analytics reports available in the system.
 *
 * @enum ReportType
 * @property {string} REVENUE - Revenue analysis reports
 * @property {string} OCCUPANCY - Occupancy rate reports
 * @property {string} GUEST - Guest analytics reports
 * @property {string} PERFORMANCE - Performance metrics reports
 * @property {string} COMPETITIVE - Competitive analysis reports
 * @property {string} FORECAST - Forecasting and prediction reports
 * @property {string} CUSTOM - Custom report types
 */
/**
 * Analytics Type Definitions for Buffr Host Hospitality Platform
 * @fileoverview Analytics type definitions for business intelligence, reporting, and data analytics operations
 * @location buffr-host/lib/types/analytics.ts
 * @purpose analytics type definitions for business intelligence, reporting, and data analytics operations
 * @modularity Centralized type definitions providing type safety across the entire Buffr Host application
 * @api_integration REST API endpoints, HTTP request/response handling
 * @security Type-safe security definitions for authentication, authorization, and data protection
 * @ai_integration Machine learning and AI service type definitions for predictive analytics
 * @authentication User authentication and session management type definitions
 * @typescript_strict Strict TypeScript type safety ensuring compile-time error prevention
 * @type_safety Comprehensive type coverage preventing runtime errors and improving developer experience
 * @scalability Type definitions supporting horizontal scaling and multi-tenant architecture
 * @maintainability Self-documenting types enabling easier code maintenance and refactoring
 * @testing Type-driven development supporting comprehensive test coverage
 *
 * Type Definitions Summary:
 * - 25 Interfaces: AnalyticsRequest, RevenueAnalyticsRequest, RevenueAnalyticsResponse...
 * - 3 Enums: ReportType, GroupBy, MetricType
 * - Total: 28 type definitions
 *
 * Usage and Integration:
 * - Frontend Components: Type-safe props and state management
 * - API Routes: Request/response type validation
 * - Database Services: Schema-aligned data operations
 * - Business Logic: Domain-specific type constraints
 * - Testing: Type-driven test case generation
 *
 * @example
 * // Import type definitions
 * import type { ReportType, GroupBy, MetricType... } from './analytics';
 *
 * // Usage in component
 * interface ComponentProps {
 *   data: AnalyticsRequest;
 *   onAction: (event: EventType) => void;
 * }
 *
 * @example
 * // Database service usage
 * const userService = {
 *   async getUser(id: string): Promise<User> {
 *     // Type-safe database operations
 *     return await db.query('SELECT * FROM users WHERE id = $1', [id]);
 *   }
 * };
 *
 * Exported Types:
 * @typedef {Enum} ReportType
 * @typedef {Enum} GroupBy
 * @typedef {Enum} MetricType
 * @typedef {Interface} AnalyticsRequest
 * @typedef {Interface} RevenueAnalyticsRequest
 * @typedef {Interface} RevenueAnalyticsResponse
 * @typedef {Interface} OccupancyAnalyticsRequest
 * @typedef {Interface} OccupancyAnalyticsResponse
 * @typedef {Interface} GuestAnalyticsRequest
 * @typedef {Interface} GuestAnalyticsResponse
 * ... and 18 more type definitions
 *
 * @returns {Object} Type definitions module with all exported types and interfaces
 * @type_safety Ensures 100% type coverage across the Buffr Host application
 * @documentation Comprehensive JSDoc documentation for developer experience
 * @maintainability Type-driven development enabling safe refactoring and evolution
 */

export enum ReportType {
  REVENUE = 'revenue',
  OCCUPANCY = 'occupancy',
  GUEST = 'guest',
  PERFORMANCE = 'performance',
  COMPETITIVE = 'competitive',
  FORECAST = 'forecast',
  CUSTOM = 'custom',
}

/**
 * Group By Enumeration
 *
 * Defines the time periods for grouping analytics data.
 *
 * @enum GroupBy
 * @property {string} DAY - Group data by day
 * @property {string} WEEK - Group data by week
 * @property {string} MONTH - Group data by month
 * @property {string} QUARTER - Group data by quarter
 * @property {string} YEAR - Group data by year
 */
export enum GroupBy {
  DAY = 'day',
  WEEK = 'week',
  MONTH = 'month',
  QUARTER = 'quarter',
  YEAR = 'year',
}

/**
 * Metric Type Enumeration
 *
 * Defines the types of metrics that can be analyzed in analytics reports.
 *
 * @enum MetricType
 * @property {string} REVENUE - Total revenue metrics
 * @property {string} OCCUPANCY - Occupancy rate metrics
 * @property {string} ADR - Average Daily Rate (for hotels)
 * @property {string} REVPAR - Revenue Per Available Room (for hotels)
 * @property {string} GUEST_COUNT - Number of guests
 * @property {string} BOOKING_COUNT - Number of bookings
 * @property {string} SATISFACTION - Customer satisfaction scores
 */
export enum MetricType {
  REVENUE = 'revenue',
  OCCUPANCY = 'occupancy',
  ADR = 'adr',
  REVPAR = 'revpar',
  GUEST_COUNT = 'guest_count',
  BOOKING_COUNT = 'booking_count',
  SATISFACTION = 'satisfaction',
}

/**
 * Analytics Request Interface
 *
 * Base interface for all analytics request types. Contains common fields required
 * for analytics queries.
 *
 * @interface AnalyticsRequest
 * @property {string} property_id - Property ID to analyze (required)
 * @property {Date} start_date - Start date for the analysis period
 * @property {Date} end_date - End date for the analysis period
 * @property {GroupBy} group_by - Time period grouping for the data
 *
 * @example
 * const request: AnalyticsRequest = {
 *   property_id: 'prop_123',
 *   start_date: new Date('2024-01-01'),
 *   end_date: new Date('2024-01-31'),
 *   group_by: GroupBy.DAY
 * };
 */
export interface AnalyticsRequest {
  property_id: string;
  start_date: Date;
  end_date: Date;
  group_by: GroupBy;
}

export interface RevenueAnalyticsRequest extends AnalyticsRequest {
  include_breakdown: boolean;
  include_growth: boolean;
}

export interface RevenueAnalyticsResponse {
  property_id: string;
  period: Record<string, string>;
  summary: Record<string, number>;
  revenue_trend: Array<Record<string, any>>;
  room_type_breakdown: Array<Record<string, any>>;
  generated_at: string;
}

export interface OccupancyAnalyticsRequest extends AnalyticsRequest {
  include_room_types: boolean;
  include_seasonality: boolean;
}

export interface OccupancyAnalyticsResponse {
  property_id: string;
  period: Record<string, string>;
  summary: Record<string, number>;
  daily_occupancy: Array<Record<string, any>>;
  generated_at: string;
}

export interface GuestAnalyticsRequest {
  property_id: string;
  days: number;
  include_demographics: boolean;
  include_behavior: boolean;
}

export interface GuestAnalyticsResponse {
  property_id: string;
  period_days: number;
  summary: Record<string, any>;
  demographics: Record<string, any>;
  generated_at: string;
}

export interface PerformanceAnalyticsRequest {
  property_id: string;
  days: number;
  include_trends: boolean;
  include_forecasting: boolean;
}

export interface PerformanceAnalyticsResponse {
  property_id: string;
  property_name: string;
  period_days: number;
  kpis: Record<string, any>;
  revenue_metrics: Record<string, any>;
  occupancy_metrics: Record<string, any>;
  guest_metrics: Record<string, any>;
  market_comparison: Record<string, any>;
  trends: Record<string, any>;
  generated_at: string;
}

export interface CompetitiveAnalysisRequest {
  property_id: string;
  competitor_properties: string[];
  metrics: MetricType[];
}

export interface CompetitiveAnalysisResponse {
  property_id: string;
  competitors: number;
  analysis: Record<string, any>;
  generated_at: string;
}

export interface ForecastingRequest {
  property_id: string;
  forecast_days: number;
  include_confidence: boolean;
  include_scenarios: boolean;
}

export interface ForecastingResponse {
  property_id: string;
  forecast_days: number;
  forecast: Record<string, any>;
  confidence_level: number;
  generated_at: string;
}

export interface CustomReportRequest {
  property_id: string;
  report_type: ReportType;
  start_date: Date;
  end_date: Date;
  metrics: MetricType[];
  filters?: Record<string, any>;
  group_by?: GroupBy;
  format: string; // json, csv, pdf
}

export interface CustomReportResponse {
  property_id: string;
  report_type: string;
  period: Record<string, string>;
  metrics: Record<string, any>;
  generated_at: string;
}

export interface KPISummary {
  total_revenue: number;
  average_occupancy: number;
  average_daily_rate: number;
  revenue_per_available_room: number;
  total_guests: number;
  repeat_guest_rate: number;
  average_stay_length: number;
  revenue_growth: number;
}

export interface TrendAnalysis {
  metric: string;
  trend_direction: string; // increasing, decreasing, stable
  change_percentage: number;
  period: string;
  confidence: number;
}

export interface MarketComparison {
  metric: string;
  property_value: number;
  market_average: number;
  variance: number;
  percentile: number;
  performance: string; // above_average, average, below_average
}

export interface DashboardData {
  property_id: string;
  property_name: string;
  current_period: Record<string, string>;
  kpis: KPISummary;
  trends: TrendAnalysis[];
  market_comparison: MarketComparison[];
  generated_at: string;
}

export interface RealTimeMetrics {
  property_id: string;
  timestamp: Date;
  occupancy_rate: number;
  revenue_today: number;
  bookings_today: number;
  cancellations_today: number;
  average_rating: number;
  active_guests: number;
  available_rooms: number;
}

export interface MLModelMetrics {
  model_name: string;
  accuracy: number;
  precision: number;
  recall: number;
  f1_score: number;
  last_trained: Date;
  predictions_today: number;
  confidence_threshold: number;
}

export interface AITaskStatus {
  task_id: string;
  status: string; // pending, running, completed, failed
  progress: number;
  result?: unknown;
  error?: string;
  created_at: Date;
  completed_at?: Date;
}

export interface DataQualityMetrics {
  property_id: string;
  data_completeness: number;
  data_accuracy: number;
  data_consistency: number;
  missing_fields: string[];
  duplicate_records: number;
  last_updated: Date;
}

export interface BusinessIntelligenceInsights {
  property_id: string;
  insight_type: string;
  title: string;
  description: string;
  impact: string; // high, medium, low
  confidence: number;
  actionable: boolean;
  recommendations: string[];
  generated_at: Date;
}

/**
 * Revenue Analytics Interface
 *
 * Main interface for revenue analytics data and operations.
 */
export interface RevenueAnalytics {
  id: string;
  propertyId: string;
  period: {
    startDate: Date;
    endDate: Date;
  };
  revenue: {
    total: number;
    breakdown: {
      bookings: number;
      services: number;
      other: number;
    };
  };
  metrics: {
    averageDailyRate: number;
    occupancyRate: number;
    revenuePerAvailableRoom: number;
  };
  trends: {
    direction: 'up' | 'down' | 'stable';
    percentage: number;
    period: string;
  };
  forecasts: {
    nextMonth: number;
    nextQuarter: number;
    confidence: number;
  };
}
